### ISSUE TITLE
borg check --repair runs very slow (and corrupts archives)

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#3444]
from @bebehei on IRC (IRC log shortened):

borg 1.1.3 on client and server, running borg check --repair on a corrupt repository hangs for >5h after emitting a msg about missing metadata.

```
23:11  bebehei$ Now it happened: Oh, the metadata chunk is missing.
23:12  bebehei$ strace output slowed down immediately
23:16  bebehei$ item metadata chunk missing [chunk: 001056_bdee87d...a3e50d]
23:16  bebehei$ But the borg serve process is dead.
23:22  ThomasWaldmann$ archive.py:1497 is the place where that msg gets emitted
```

Bounty: https://www.bountysource.com/issues/52774936-borg-check-repair-hangs-infinitely

**Comment by ThomasWaldmann** [comment]
The bounty is for the bug being found and fixed.

To prove that it is this bug and it is fixed, maybe a unit test is needed - reproducing the bug first (hanging) and then succeeding after the fix.

**Comment by ThomasWaldmann** [comment]
Hmm, I suspect this bug is even worse.

As the "hang" is not really a infinite hang (but just borg being very slow) and the bug in the validator basically lets borg check "think" every item (even good ones) is invalid once it had to switch into resync-mode, it will destroy the rest of the archive after a missing chunk in the archive metadata stream. :-(

So, if you don't kill borg check (and thus, abort the transaction), but wait patiently for borg check to finish (and the transaction gets committed), all affected archives will be severely damaged.

**Comment by ThomasWaldmann** [comment]
Fix is coming, see #3445.
